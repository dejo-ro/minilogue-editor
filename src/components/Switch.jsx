/**
 * A multi-valued toggle switch component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import './Switch.css';

export default class Switch extends React.Component {
  static propTypes = {
    value: PropTypes.number.isRequired,
    numPositions: PropTypes.number.isRequired,
    vertical: PropTypes.bool,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    vertical: true,
    onChange: () => {},
  };

  state = {
    value: this.props.value,
  }

  updateValueElement = () => {
    // Calculate the top offset
    const { height: rangeHeight, width: rangeWidth } = this.rangeEl.getBoundingClientRect();
    const { height: valueHeight, width: valueWidth } = this.valueEl.getBoundingClientRect();
    const rangeSize = this.props.vertical ? rangeHeight : rangeWidth;
    const valueSize = this.props.vertical ? valueHeight : valueWidth;
    const stepSize = (rangeSize - valueSize) / (this.props.numPositions - 1);
    const offset = this.state.value * stepSize;
    const styleProperty = this.props.vertical ? 'bottom' : 'left';
    this.valueEl.style[styleProperty] = `${offset}px`;
  }

  componentDidMount = () => this.updateValueElement();
  componentDidUpdate = () => this.updateValueElement();

  componentWillReceiveProps = (nextProps) => {
    if (this.state.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  onMouseDown = (event) => {
    this.onMouseMove(event);
    event.preventDefault();
    document.addEventListener('mousemove', this.onMouseMove);
    document.addEventListener('mouseup', this.onMouseUp);
  }

  onMouseUp = () => {
    document.removeEventListener('mousemove', this.onMouseMove);
    document.removeEventListener('mouseup', this.onMouseUp);
  }

  handleMousePosition = (event) => {
    const { clientX, clientY } = event;
    this.setState((prevState, props) => {
      const {
        height, width, bottom, left,
      } = this.rangeEl.getBoundingClientRect();
      const size = props.vertical ? height : width;
      const positionSize = size / props.numPositions;
      const relativePosition = props.vertical ? bottom - clientY : clientX - left;
      const positionIndex = Math.max(
        0,
        Math.min(
          Math.floor(relativePosition / positionSize),
          props.numPositions - 1,
        ),
      );

      props.onChange(positionIndex);
      return {
        value: positionIndex,
      };
    });
  }

  onMouseMove = event => this.handleMousePosition(event);

  onRangeClick = (event) => {
    event.preventDefault();
    this.handleMousePosition(event);
  }

  onWheel = (event) => {
    event.preventDefault();
    const { deltaY } = event;
    this.setState((prevState, props) => {
      const { height, width } = this.rangeEl.getBoundingClientRect();
      const size = props.vertical ? height : width;
      const positionSize = size / (props.numPositions - 1);
      const delta = Math.round(-deltaY / positionSize);
      const clamped = Math.max(
        0,
        Math.min(prevState.value + delta, props.numPositions - 1),
      );
      props.onChange(clamped);
      return {
        value: clamped,
      };
    });
  }

  render() {
    return (
    <div
      className={`switch-range ${this.props.vertical ? 'switch-range-vertical' : 'switch-range-horizontal'}`}
      ref={(rangeEl) => { this.rangeEl = rangeEl; }}
      onClick={this.onRangeClick}
      onWheel={this.onWheel}
    >
      <div
        className="switch-value"
        ref={(valueEl) => { this.valueEl = valueEl; }}
        onMouseDown={this.onMouseDown}
      />
    </div>
    );
  }
}
