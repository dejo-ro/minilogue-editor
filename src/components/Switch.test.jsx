import React from 'react';
import renderer from 'react-test-renderer';
import Switch from './Switch.jsx';

const createNodeMock = () => ({
  getBoundingClientRect: jest.fn(() => (
    {
      width: 120,
      height: 120,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    }
  )),
  style: {},
});

describe('switch component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Switch value={0} numPositions={5} />, { createNodeMock })
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
