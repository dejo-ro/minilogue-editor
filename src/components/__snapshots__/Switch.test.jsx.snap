// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch component renders correctly 1`] = `
<div
  className="switch-range switch-range-vertical"
  onClick={[Function]}
  onWheel={[Function]}
>
  <div
    className="switch-value"
    onMouseDown={[Function]}
  />
</div>
`;
